ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PackageCommitter.scala

```diff
@@ -205,52 +134,314 @@ private[kvutils] class PackageCommitter(
-            .consume(
-              _ => sys.error("Unexpected request to PCS in preloadPackage"),
-              pkgId => packages.get(pkgId),
-              _ => sys.error("Unexpected request to keys in preloadPackage")
-            )
+  final protected val enqueuePreload: Step = {
+    case (_, partialResult @ (uploadEntry, _)) =>
+      preloadExecutor.execute { () =>
+        traceLog(s"Uploading ${uploadEntry.getArchivesCount} archive", uploadEntry)
+        decodePackages(uploadEntry)
+          .flatMap(uploadPackages)
+          .fold(
+            msg => traceLog(s"Uploading failed: $msg", uploadEntry),
+            _ => traceLog(s"Uploading successful", uploadEntry),
+          )
+        metrics.daml.kvutils.committer.packageUpload.loadedPackages(() =>
+          engine.compiledPackages().packageIds.size)
       }
-      trace("Preload complete")
-    } catch {
-      case scala.util.control.NonFatal(err) =>
-        logger.error(
-          s"Preload exception, correlationId=$submissionId error='$err' stackTrace='${err.getStackTrace
-            .mkString(", ")}'")
-    } finally {
-      val _ = ctx.stop()
-    }
+      StepContinue(partialResult)
+  }
+
+  // Filter out packages already on the ledger.
+  // Should be done after decoding, validation or preload, as those step may
+  // require packages on the ledger by not loaded by the engine.
+  protected val filterKnownPackages: Step = {
+    case (ctx, (uploadEntry, pkgs)) =>
+      val archives = uploadEntry.getArchivesList.asScala.filter { archive =>
+        val stateKey = DamlStateKey.newBuilder
+          .setPackageId(archive.getHash)
+          .build
+        ctx.get(stateKey).isEmpty
+      }
+      StepContinue(uploadEntry.clearArchives().addAllArchives(archives.asJava) -> pkgs)
+  }
+
+  final private[committer] val buildLogEntry: Step = {
+    case (ctx, (uploadEntry, _)) =>
+      metrics.daml.kvutils.committer.packageUpload.accepts.inc()
+      logger.trace(
+        s"Packages committed, packages=[${uploadEntry.getArchivesList.asScala.map(_.getHash).mkString(", ")}] correlationId=${uploadEntry.getSubmissionId}")
+
+      uploadEntry.getArchivesList.forEach { archive =>
+        ctx.set(
+          DamlStateKey.newBuilder.setPackageId(archive.getHash).build,
+          DamlStateValue.newBuilder.setArchive(archive).build
+        )
+      }
+      ctx.set(
+        packageUploadDedupKey(ctx.getParticipantId, uploadEntry.getSubmissionId),
+        DamlStateValue.newBuilder
+          .setSubmissionDedup(DamlSubmissionDedupValue.newBuilder)
+          .build
+      )
+      val successLogEntry =
+        buildLogEntryWithOptionalRecordTime(ctx.getRecordTime, _.setPackageUploadEntry(uploadEntry))
+      if (ctx.preExecute) {
+        setOutOfTimeBoundsLogEntry(uploadEntry, ctx)
+      }
+      StepStop(successLogEntry)
   }
 
+  final override protected val committerName: String = "package_upload"
+
+  protected def specific_steps: List[(StepInfo, Step)]
```

Comment:
I'm not sure about the name, but let's use regular scala naming conventions ;)
```suggestion
  protected def specificSteps: List[(StepInfo, Step)]
```